### ISSUE TITLE
Replace Pool CRD with a PoolTemplate CRD

### ISSUE THREAD

**Issue by travisn** [rook/rook#1109]
To improve the design of pools and their usage from storage classes, object stores, and file systems, the design question was raised about whether the Pool CRD should actually create pools, or whether it should just be a template for creating pools. 

This topic is related to the point of discussion around the file and object stores that consume multiple pools. The pools must be created with certain names and properties and the lifecycle of the pools should be owned by the file or object store. In the current implementation, the pool properties were defined inline with those CRDs. However, that leads to duplication of properties of the same type of pool. A better approach would be that we define a template for the pool settings as a CRD. Then in the file and object CRDs, you would just need to refer to that template by name.

To define a PoolTemplate CRD, this would mean that:
- Pools do not have a lifecycle of their own. They belong to another resource such as the block, file, or object store. 
- You want to define the properties of a pool in a central place without duplicating those settings across all of the block, file, and object consumers. For example, I should only need to define the pool settings I care about.

As an example, perhaps in my cluster I want all replicated pools to have replication=3 and erasure coding with k=6 and m=4. Thus, I would create just two pool templates and all the block, file, and object stores could refer to them. Or I could get as granular as needed and define different pool settings for every store.

The schema for the PoolTemplate CRD would be the same as Pool. 
- The `StorageClass` would refer to the name of the pool template
- The `ObjectStore` and `Filesystem` CRD would refer to the name of the metadata and data pool templates.

In fact, we don't need the Pool type anymore as long as there is not a scenario for the pools to have a lifecycle of their own.

For an example, let's define two pool templates, one for replication and one for erasure coding.
```
apiVersion: rook.io/v1alpha1
kind: PoolTemplate
metadata:
  name: my-replica-pool
  namespace: rook
spec:
  replicated:
    size: 3
---
apiVersion: rook.io/v1alpha1
kind: PoolTemplate
metadata:
  name: my-ec-pool
  namespace: rook
spec:
  failureDomain: osd
  erasureCoded:
    dataChunks: 4
    codingChunks: 2
```
An example of the Filesystem CRD would be:
```
apiVersion: rook.io/v1alpha1
kind: Filesystem
metadata:
  name: myfs
  namespace: rook
spec:
  metadataPool: my-replica-pool
  dataPools:
    - my-ec-pool
  metadataServer:
    activeCount: 1
    activeStandby: true
```
And the ObjectStore CRD would be:
```
apiVersion: rook.io/v1alpha1
# For earlier versions than K8s 1.7, the kind must be Objectstore
kind: ObjectStore
metadata:
  name: my-store
  namespace: rook
spec:
  metadataPool: my-replica-pool
  dataPool: my-ec-pool
  gateway:
    type: s3
    port: 80
    instances: 1
```

I would recommend we get this into 0.6 since it will be more difficult to change after pools are in beta.

@bassam @jbw976 @kokhang  thoughts on this?

**Comment by jbw976** [comment]
If a StorageClass is simply referring to a pool template, what does that mean for pool reuse?  Two examples I can think of:
1. Two PVC's use the same StorageClass.  The first PVC has the pool automatically created.  Does the second PVC see that the pool already exists and reuses it?
1. Two StorageClasses use the same PoolTemplate.  Does each StorageClass create independent pools with the identical properties?  Or do they both use the same underlying pool?

How would a user create a pool that is not intended for use by an explicit instance of block, file, or object? I have a feeling that someone would still want to be able to do that.

**Comment by travisn** [comment]
Good questions, the lifecycle doesn't seem to make sense with the pool inside a storage class as your first question indicates. Also, if you delete a storage class, our code wouldn't be triggered in order to delete the pool. 

For the second question, I would expect two different pools to be created. For the last question, you could still use the ceph tools to create a pool if you have some unexpected scenario.

But I'm leaning towards not changing anything for the block scenarios and leaving the Pool CRD. We could still add the PoolTemplate CRD to be used by file and object, but it wouldn't impact block.

**Comment by bassam** [comment]
this is a tricky one. I think its worth having a longer design discussion on our use of CRDs, validation, reuse, dependencies, status, etc. SIG-API is looking to invest here and I think we should figure out what improvements we'd like to make.

**Comment by jbguerraz** [comment]
- Two PVC's use the same StorageClass. The first PVC has the pool automatically created. Does the second PVC see that the pool already exists and reuses it?
  - I'd say yes (as the relations are currently : pool -1-1-> storageclass -1-n-> pvc)
- Two StorageClasses use the same PoolTemplate. Does each StorageClass create independent pools with the identical properties? Or do they both use the same underlying pool?
  - 2 pools with the same properties would be expected IMHO

It could use new "metadataPoolTemplate", "dataPoolTemplate", "poolTemplate" properties and keep the current "metadataPool", "dataPool", "pool" properties.
Then if for instance "metadataPoolTemplate" is defined rook would create a pool based on that template and set the "metadataPool" property accordingly.
That means if one need, one could still use the "pool" CRD and set the "metadataPool" property just as now (it would then ignore the "metadataPoolTemplate" property)
Basically the "poolTemplate" CRD will be a way to get rook to create (and update/delete : maintain) "pool" resources automatically and link them up to the resources that needs it.
